lib/assembly_report/assembly_report.hpp

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ASSEMBLY_REPORT_HPP
+#define ASSEMBLY_REPORT_HPP
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include <boost/algorithm/string/trim.hpp>
+#include "util/stream_utils.hpp"
+#include "util/string_utils.hpp"
+
+namespace ebi
+{
+  namespace assembly_report
+  {
+    /*
+     * Struct to contain all the synonyms of a contig
+     */
+    struct Synonyms_list {
+        std::vector<std::string> list;
+        Synonyms_list() {
+
+        }
+
+        Synonyms_list(std::string contig) {
+            list.push_back(contig);
+        }
+
+        Synonyms_list(std::vector<std::string> & list) : list(list) {
+
+        }
+    };
+
+    /*
+     * Map to map all contigs to their synonym_maps
+     */
+    class Synonyms_map {
+        std::map<std::string,int> index_map;
+        std::map<int,Synonyms_list> list_map;
+
+        Synonyms_list extract_synonyms(std::vector<std::string> & list) {
+            std::vector<int> synonym_indices{0,4,6,9};
+            Synonyms_list synonyms_list;
+            for (auto index : synonym_indices) {
+                synonyms_list.list.push_back(list[index]);
+            }
+
+            return synonyms_list;
+        }
+
+        public:
+        Synonyms_list get_synonym_list(std::string & contig) {
+            if (index_map.find(contig) != index_map.end()) {
+                return list_map[index_map[contig]];
+            }
+            else {
+                return Synonyms_list(contig);
+            }
+        }
+
+        void parse_file(std::istream &report) {
+            std::vector<char> report_line;
+            size_t const default_line_buffer_size = 64 * 1024;
+            size_t const assembly_report_line_size = 10;
+            report_line.reserve(default_line_buffer_size);
+
+            int list_index = 0;
+            while (util::readline(report, report_line).size() != 0) {
+                if (report_line[0] == '#') {
+                    continue;
+                }
+
+                std::vector<std::string> list;
+                std::string line{report_line.begin(), report_line.end()};
+                boost::algorithm::trim(line);
+                util::string_split(line, "\t", list);
+                if(list.size() != assembly_report_line_size) {
+                    // report wrong line
+                    continue;
+                }
+
+                auto synonyms_list = extract_synonyms(list);
+                list_map[list_index] = synonyms_list;
+                for (auto synonym : synonyms_list.list) {
+                    index_map[synonym] = list_index;
+                }
+                list_index++;
```

Comment:
@jmmut would it be fine for keeping its name as `list_map` as it is. even after changing the datatype from map to vector ? I think this name will be fine as it is still working as a map mapping from index to list.

Reply:
list_map doesn't really say anything to me. I just spent two minutes trying to remember what it holds. Variable names are exactly meant to minimize this time, "context-less code reading" is what you are trying to optimize with variable names. When you are developing the code you are not affected by this delay because it's already in your working memory. 

We basically have a list of lists.
```
contig_1_a, contig_1_b, contig_1_c
contig_2_a, ...
...
```
so I would call each cell a `synonym` or `contig_synonym`, each row `synonyms` or `contig_synonyms` (or `synonyms_list` but "list" is not adding information and making it dependent on implementation and confusing if refactored) , and the whole set of rows are `contigs`

Reply:
and it's not working as a map mapping from index to list because the index is an implementation detail, and the user doesn't know about those indexes, and not even internally you have an index that represents any information and you want the mapping of that information. the index is just a way to loop over the elements.